services-core/src/solution_submission/first_match.rs

```diff
@@ -0,0 +1,102 @@
+use futures::{
+    future::FusedFuture,
+    stream::{futures_unordered::FuturesUnordered, StreamExt as _},
+};
+use std::{future::Future, pin::Pin, task::Context, task::Poll};
+
+/// Resolves to the first result of a set of inner futures that matches a filter.
+/// Futures can be dynamically added to the inner future.
+/// Stays pending until there is at least one future.
+/// If the last inner future is ready its result is returned even if it doesn't match the filter.
```

Comment:
We do need this last logic future for the way we use this struct. If we wanted the distinction through returning an enum then it would probably make the code using this struct more complicated. I will do the rename.